data/txHandler.go

```diff
@@ -529,6 +558,24 @@ func (handler *TxHandler) processIncomingTxn(rawmsg network.IncomingMessage) net
 	dec := protocol.NewMsgpDecoderBytes(rawmsg.Data)
 	ntx := 0
 	consumed := 0
+
+	var err error
+	var capguard *util.ErlCapacityGuard
+	if handler.erl != nil {
+		// consume a capacity unit
+		capguard, err = handler.erl.ConsumeCapacity(rawmsg.Sender.(util.ErlClient))
+		if err != nil {
+			handler.erl.EnableCongestionControl()
```

Comment:
It looks strange to read that an error happened, and the response is to enable something.
It seems that the err  is not necessarily an unexpected situation, but is a control parameter. ConsumeCapacity does not describe in the comments what errors are returned, and need digging into the internal functions to understand what is expected.
Either add a comment to describe what err entails, or use a different control parameter.

Reply:
That's a fair point -- using an error as a control is uncommon. But I think it makes sense here, so it does deserve a comment.

The current strategy for TX Handler to use the ERL is that it asks for capacity, and while it is available, everything is fine. However, if there should ever be an *issue granting capacity*, then TX Handler decides that the ERL should use congestion management.

That "Issue Granting Capacity" indicates that at the moment, the underlying TX Backlog is full. It is indeed an error, but it's an error we encounter any time the inputs overwhelm this handler, so it's not exactly "unexpected" like you mention. Pre-ERL this error was/is just as likely, but it happens lower down, when the TXHandler attempts to put the item onto the backlog and fails (and no sort of function calling happens inside the condition).

Another way of putting this -- enablement of congestion control is indeed the response to an error condition -- if the handler is being overwhelmed, that's an error. And if it's generating errors for being overwhelmed, that's our signal to turn on defensive measures (congestion management)

Reply:
> That's a fair point -- using an error as a control is uncommon. But I think it makes sense here, so it does deserve a comment.

That is all I need :-)